hardware_interface/mainpage.dox

```diff
@@ -0,0 +1,303 @@
+  hardware interface to support reading the state of an array of named actuators,
+  each of which has some position, velocity, and effort (force or torque).
+- [ActuatorCommandInterfaces](\ref hardware_interface::ActuatorCommandInterface)
+  - [EffortActuatorInterface](\ref hardware_interface::EffortActuatorInterface)
+  - [VelocityActuatorInterface](\ref hardware_interface::VelocityActuatorInterface)
+  - [PositionActuatorInterface](\ref hardware_interface::PositionActuatorInterface)
+- [PosVelJointInterface](\ref hardware_interface::PosVelJointInterface)
+- [PosVelAccJointInterface](\ref hardware_interface::PosVelAccJointInterface)
+- [Force-torque sensor Interface](\ref hardware_interface::ForceTorqueSensorInterface)
+- [IMU sensor Interface](\ref hardware_interface::ImuSensorInterface)
+
+Note that \ref hardware_interface::JointCommandInterface allows both reading
+joint state and commanding [effort|velocity|position]-based joints
+(see this [answer](https://answers.ros.org/question/209619/differences-between-hardware-interfaces/?answer=209636#post-id-209636)).
+
+\section new-robot-setup Setting up a new robot
+
+The following example explains how to set up a new robot to work with the
+[controller_manager](http://wiki.ros.org/controller_manager). When you make your
+robot support one or more of the standard interfaces, you will be able to take
+advantage of a large library of controllers (see
+[ros_controllers](https://github.com/ros-controls/ros_controllers))
+that work on the standard interfaces mentioned above.
+
+![hw_interface_2.png](https://raw.githubusercontent.com/wiki/ros-controls/ros_control/hw_interface_2.png)
+
+As the image above suggests, a custom robot hardware interface is not
+limited to be composed of only one single interface. The robot can provide as
+many interfaces as required. It could for example provide both the
+\ref hardware_interface::PositionJointInterface and the
+\ref hardware_interface::VelocityJointInterface, and many more.
+
+Assuming a robot with two joints: A & B where both joints are position controlled.
+Such a robot is defined with a class derived from \ref hardware_interface::RobotHW
+that should provide the standard [PositionJointInterface](\ref hardware_interface::PositionJointInterface) and the
+[JointStateInterface](\ref hardware_interface::JointStateInterface), so it can re-use all controllers that are already
+written to work with the [PositionJointInterface](\ref hardware_interface::PositionJointInterface) and the
+[JointStateInterface](\ref hardware_interface::JointStateInterface). The code would look like this:
+
+\code
+#include <hardware_interface/joint_command_interface.h>
+#include <hardware_interface/joint_state_interface.h>
+#include <hardware_interface/robot_hw.h>
+
+class MyRobot : public hardware_interface::RobotHW
+{
+public:
+  MyRobot() 
+  { 
+    // Initialization of the robot's resources (joints, sensors, actuators) and
+    // interfaces can be done here or inside init().
```

Comment:
Maybe we could add here that it is possible to get the joint names from the URDF via the parameter server like the [ros_control_boilerplate/generic_hw_interface](https://github.com/PickNikRobotics/ros_control_boilerplate/blob/5e40c7e875f7ed58ed061f37d4898865e382170e/src/generic_hw_interface.cpp#L52) seems to do it (using [rosparam_shortcuts](https://github.com/PickNikRobotics/rosparam_shortcuts)).

Reply:
```suggestion
    // Initialization of the robot's resources (joints, sensors, actuators) and
    // interfaces can be done here or inside init().
    // E.g. parse the URDF for joint names & interfaces, then initialize them
```